src/Http/WebUtilities/src/AspNetCoreTempDirectory.cs

```diff
@@ -0,0 +1,37 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.IO;
+
+namespace Microsoft.AspNetCore.Internal
+{
+    internal static class AspNetCoreTempDirectory
+    {
+        private static string _tempDirectory;
+
+        public static string TempDirectory
+        {
+            get
+            {
+                if (_tempDirectory == null)
+                {
+                    // Look for folders in the following order.
+                    var temp = Environment.GetEnvironmentVariable("ASPNETCORE_TEMP") ?? // ASPNETCORE_TEMP - User set temporary location.
```

Comment:
I didn't look into this in great detail. Any suggestions on how this could be done without making the new Stream gross?

Reply:
It would be a new options type such as TempDirectoryOptions.

Reply:
If these types required the option type to be specified to construct, I can see the value in having these. Requiring the caller to get the path to the temp directory from an option type before passing it in to these types seems onerous to the caller and super easy to get wrong. Is there value in doing this? Additionally, is this a common requirement for users to want to customize where the ASP.NET Core specific temp directory is located?

Reply:
This is one of the few parts of the stack that is only configurable via environment variable. Yes people have asked to change it. It makes sense to tie it into our config system like everything else. If you don't want to deal with the options directly then maybe FileBufferingReadStream should be produced from a factory/service.

Reply:
Filed https://github.com/aspnet/AspNetCore/issues/9466 so we can track this separately. Changing the read stream would be a breaking change and I wanted to avoid conflating it with this PR